beacon-chain/endtoend/endtoend_test.go

```diff
@@ -0,0 +1,514 @@
+	tmpPath := config.tmpPath
+	index := len(beaconNodes)
+	binaryPath, found := bazel.FindBinary("beacon-chain", "beacon-chain")
+	if !found {
+		t.Log(binaryPath)
+		t.Fatal("beacon chain binary not found")
+	}
+	file, err := os.Create(path.Join(tmpPath, fmt.Sprintf("beacon-%d.log", index)))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	args := []string{
+		"--no-discovery",
+		"--http-web3provider=http://127.0.0.1:8545",
+		"--web3provider=ws://127.0.0.1:8546",
+		"--p2p-host-ip=127.0.0.1",
+		fmt.Sprintf("--datadir=%s/eth2-beacon-node-%d", tmpPath, index),
+		fmt.Sprintf("--deposit-contract=%s", config.contractAddr.Hex()),
+		fmt.Sprintf("--rpc-port=%d", 4000+index),
+		fmt.Sprintf("--p2p-tcp-port=%d", 13000+index),
+		fmt.Sprintf("--monitoring-port=%d", 8080+index),
+		fmt.Sprintf("--grpc-gateway-port=%d", 3200+index),
+	}
+
+	if config.minimalConfig {
+		args = append(args, "--minimal-config")
+	}
+	// After the first node is made, have all following nodes connect to all previously made nodes.
+	if index >= 1 {
+		for p := 0; p < index; p++ {
+			args = append(args, fmt.Sprintf("--peer=%s", beaconNodes[p].multiAddr))
+		}
+	}
+
+	cmd := exec.Command(binaryPath, args...)
+	cmd.Stderr = file
+	cmd.Stdout = file
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("failed to start beacon node: %v", err)
+	}
+
+	if err = WaitForTextInFile(file, "Connected to eth1 proof-of-work"); err != nil {
+		t.Fatal(err)
+	}
+	t.Logf("Beacon node %d started", index)
+
+	response, err := http.Get(fmt.Sprintf("http://127.0.0.1:%d/p2p", 8080+index))
+	if err != nil {
+		t.Fatalf("failed to get p2p info: %v", err)
+	}
+	time.Sleep(1 * time.Second)
+
+	// Get the response body as a string
+	dataInBytes, err := ioutil.ReadAll(response.Body)
+	pageContent := string(dataInBytes)
+	if err := response.Body.Close(); err != nil {
+		t.Fatal(err)
+	}
+
+	startIdx := strings.Index(pageContent, "self=") + 5
+	multiAddr := pageContent[startIdx : startIdx+85]
+
+	return &beaconNodeInfo{
+		processID:   cmd.Process.Pid,
+		datadir:     fmt.Sprintf("%s/eth2-beacon-node-%d", tmpPath, index),
+		rpcPort:     (4000) + uint64(index),
+		monitorPort: 8080 + uint64(index),
+		grpcPort:    3200 + uint64(index),
+		multiAddr:   multiAddr,
+	}
+}
+
+// initializeValidators sends the deposits to the eth1 chain and starts the validator clients.
+func initializeValidators(
+	t *testing.T,
+	config *end2EndConfig,
+	keystorePath string,
+	beaconNodes []*beaconNodeInfo,
+) []*validatorClientInfo {
+	tmpPath := config.tmpPath
+	contractAddress := config.contractAddr
+	validatorNum := config.numValidators
+	beaconNodeNum := config.numBeaconNodes
+	binaryPath, found := bazel.FindBinary("validator", "validator")
+	if !found {
+		t.Fatal("validator binary not found")
+	}
+
+	if validatorNum%beaconNodeNum != 0 {
+		t.Fatal("Validator count is not easily divisible by beacon node count.")
+	}
+
+	valClients := make([]*validatorClientInfo, beaconNodeNum)
+	validatorsPerNode := validatorNum / beaconNodeNum
```

Comment:
Rounding? what happens when you want 5 validators and 3 beacon nodes?

Reply:
I restrict it to being cleanly divisible, should I change remove that restriction and change the math?